web/src/reducers/budget.js

```diff
@@ -497,6 +497,89 @@ const buildBudget = bigState => {
+  // basis for computing quarterly federal share.  Then reudce that to an
+  // object whose keys are FFYs and whose values are an array of percentages
+  // for fiscal quarters.
+  const percentPerQuarter = Object.entries(
+    Object.values(bigState.activities.byKey).filter(
+      a => a.name === 'Program Administration'
+    )[0].quarterlyFFP
+  ).reduce(
+    (percentYear, [ffy, quarters]) => ({
+      ...percentYear,
+      [ffy]: Object.entries(quarters)
+        .sort(([q1], [q2]) => +q1 - +q2)
+        .map(([, q]) => q.state / 100)
+    }),
+    {}
+  );
+
+  /**
+   * Update a budget section with key personnel cost numbers
+   * @param {Object} section The section to be updated; should include
+   *                 top-level properties corresponding to fiscal years along
+   *                 with a total.  Those objects have federal, state, and
+   *                 total properties.
+   * @param {Number} year The federal fiscal year to update
+   * @param {Number} federal Federal share of the cost to add
+   * @param {Number} state State share of the cost to add
+   * @param {Number} total Total cost to add
+   */
+  const updateSectionWithKP = (section, year, federal, state, total) => {
+    section[year].federal += federal;
+    section[year].state += state;
+    section[year].total += total;
+    section.total.federal += federal;
+    section.total.state += state;
+    section.total.total += total;
+  };
+
+  // We also have to account for the cost attributable to key personnel.  This
+  // isn't added to any activity, but instead gets rolled up into the totals.
+  // It is always a 90% federal/10% state split.  We're just going to make sure
+  // we only look at key personnel for whom the "has costs" flag is set, just
+  // to be safe.
+  keyPersonnel
+    .filter(kp => kp.hasCosts)
+    .forEach(kp => {
+      Object.entries(kp.costs).forEach(([year, cost]) => {
+        // Get the state and federal share.  Always 90/10 (for now).
+        const [federalShare, stateShare] = roundedPercents(cost, [0.9, 0.1]);
+        const costInfo = [year, federalShare, stateShare, cost];
+
+        // Update the various budget sections that should include key personnel
+        // costs.  :all-the-things:
+        updateSectionWithKP(newState.combined, ...costInfo);
+        updateSectionWithKP(newState.hit.statePersonnel, ...costInfo);
+        updateSectionWithKP(newState.hit.combined, ...costInfo);
+        updateSectionWithKP(newState.hitAndHie.statePersonnel, ...costInfo);
+        updateSectionWithKP(newState.hitAndHie.combined, ...costInfo);
+
+        // Compute how much of the key personnel cost is attributable to each
+        // quarter of this FFY
+        const quarterlyCosts = roundedPercents(
+          federalShare,
+          percentPerQuarter[year]
+        );
+
+        const ffyQuarterly = newState.federalShareByFFYQuarter.hitAndHie[year];
+        const ffyTotal = newState.federalShareByFFYQuarter.hitAndHie.total;
```

Comment:
The `newState` / `bigState` terminology is a little opaque to me -- I see `newState` is built from `bigState` in the reducer, but I'm having a bit of a hard time following. (I say this just because while this code seems fine in isolation, I'm not able to make sure it makes sense in context.)

Reply:
That's a good point, and the distinction is only ever implied.  The `UPDATE_BUDGET` reducer expects the entire redux state passed in with the action.  It then passes that full state object into this method, which calls it `bigState`.  `newState` is the budget component of the redux state that is being constructed and will be returned by the reducer.  Do you think calling it `newBudgetState` or something similar would help clarify it?